demo_apps/Azure_API_example/azure_api_example.ipynb

```diff
@@ -0,0 +1,610 @@
+    "from langchain.memory import ConversationBufferWindowMemory\n",
+    "\n",
+    "import langchain\n",
+    "from typing import Dict\n",
+    "import json\n",
+    "\n",
+    "langchain.debug=True\n",
+    "\n",
+    "class AzureLlamaAPIContentFormatter(ContentFormatterBase):\n",
+    "#Content formatter for Llama 2 API for Azure MaaS\n",
+    "\n",
+    "    def format_request_payload(self, prompt: str, model_kwargs: Dict) -> bytes:\n",
+    "        #Formats the request according to the chosen api\n",
+    "        prompt = ContentFormatterBase.escape_special_characters(prompt)\n",
+    "\n",
+    "        #Note how we instructed the model with system prompts. Past conversation can be past as in system prompt as well\n",
+    "        request_payload_dict = {\n",
+    "                \"messages\": [\n",
+    "                    {\"role\":\"system\", \"content\":\"The following is a conversation between a user and you. Answer the user question based on the conversation. Provide your answer only\"},\n",
+    "                    {\"role\":\"user\", \"content\":f\"{prompt}\"}\n",
+    "                    ]               \n",
+    "            }\n",
+    "        request_payload_dict.update(model_kwargs)\n",
+    "        request_payload = json.dumps(request_payload_dict)\n",
+    "        return str.encode(request_payload)\n",
+    "\n",
+    "    def format_response_payload(self, output: bytes) -> str:\n",
+    "        #Formats response\n",
+    "        return json.loads(output)[\"choices\"][0][\"message\"][\"content\"]\n",
+    "\n",
+    "#Create content fomartter\n",
+    "content_formatter = AzureLlamaAPIContentFormatter()\n",
+    "\n",
+    "#Create llm instance\n",
+    "llm = AzureMLOnlineEndpoint(\n",
+    "    endpoint_api_key=\"your-auth-key\",\n",
+    "    endpoint_url=\"https://your-endpoint.inference.ai.azure.com/v1/chat/completions\",\n",
+    "    model_kwargs={\"temperature\": 0.6, \"max_tokens\": 128, \"top_p\": 0.9},\n",
+    "    content_formatter=content_formatter,\n",
+    ")\n",
+    "\n",
+    "#Create memory\n",
+    "memory = ConversationBufferWindowMemory(llm=llm, k=5, memory_key=\"chat_history\", ai_prefix=\"Assistant\", human_prefix=\"User\")\n",
+    "\n",
+    "#Create input prompt template with chat history for chaining\n",
+    "INPUT_TEMPLATE = \"\"\"Current conversation:\n",
+    "{chat_history}\n",
+    "\n",
+    "User question:{input}\"\"\"\n",
+    "\n",
+    "conversation_prompt_template = PromptTemplate(\n",
+    "    input_variables=[\"chat_history\", \"input\"], template=INPUT_TEMPLATE\n",
+    ")\n",
+    "\n",
+    "conversation_chain_with_memory = ConversationChain(\n",
+    "    llm = llm,\n",
+    "    prompt = conversation_prompt_template,\n",
+    "    verbose = True,\n",
+    "    memory = memory,\n",
+    ")\n",
+    "\n",
+    "#Prediction\n",
+    "def predict(message, history):\n",
+    "    history_format = []\n",
```

Comment:
Looks like the code snippet below can be removed without affecting the behavior?

```
    history_format = []
    for user, assistant in history:
        history_format.append({"role": "user", "content": user })
        history_format.append({"role": "assistant", "content":assistant})
    history_format.append({"role": "user", "content": message})
```

Reply:
This is for logging the conversation history. Also it is part of Gradio's tutorial for best practice, so I will leave it here.